sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala

```diff
@@ -194,9 +194,14 @@ case class ColumnarToRowExec(child: SparkPlan) extends ColumnarToRowTransition w
        |    $shouldStop
        |  }
        |  $idx = $numRows;
+       |  $batch.closeIfNotWritable();
```

Comment:
For writable column vectors, they are reused across batches, so we cannot close them until finishing all batches (see below).